### PR TITLE
C++: Model that string functions read their buffer

### PR DESCRIPTION
**Pull request by jbj**
This fixes FNs on the `CWE-190/semmle/tainted/ArithmeticTainted.expected` test with all open PRs applied.

It turns out that any function implementing `SideEffectFunction` [is assumed not to read](https://github.com/Semmle/ql/blob/b3eada829a27e3d62d0ce4aa3e615107938cdcb6/cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedCall.qll#L545-L557) through its parameters unless `hasSpecificReadSideEffect` is implemented.

### CODE REVIEW

**Comment by jbj** [comment]
This is the diff for what's fixed:

```diff
--- expected
+++ actual
@@ -1,6 +1,9 @@
 | test3.c:15:10:15:10 | x | $@ flows to here and is used in arithmetic, potentially causing an overflow. | test3.c:11:15:11:18 | argv | User-provided value |
 | test3.c:15:14:15:14 | y | $@ flows to here and is used in arithmetic, potentially causing an overflow. | test3.c:11:15:11:18 | argv | User-provided value |
 | test3.c:15:18:15:18 | z | $@ flows to here and is used in arithmetic, potentially causing an overflow. | test3.c:11:15:11:18 | argv | User-provided value |
+| test5.cpp:17:6:17:18 | call to getTaintedInt | $@ flows to here and is used in arithmetic, potentially causing an overflow. | test5.cpp:9:7:9:9 | buf | User-provided value |
+| test5.cpp:19:6:19:6 | y | $@ flows to here and is used in arithmetic, potentially causing an overflow. | test5.cpp:9:7:9:9 | buf | User-provided value |
+| test5.cpp:19:6:19:6 | y | $@ flows to here and is used in arithmetic, potentially causing an underflow. | test5.cpp:9:7:9:9 | buf | User-provided value |
 | test.c:14:15:14:28 | maxConnections | $@ flows to here and is used in arithmetic, potentially causing an overflow. | test.c:11:29:11:32 | argv | User-provided value |
 | test.c:14:15:14:28 | maxConnections | $@ flows to here and is used in arithmetic, potentially causing an underflow. | test.c:11:29:11:32 | argv | User-provided value |
 | test.c:44:7:44:10 | len2 | $@ flows to here and is used in arithmetic, potentially causing an underflow. | test.c:41:17:41:20 | argv | User-provided value |
```

The problem the was lack of a read side effect node on the call to `strtoul`.

**Comment by rdmarsh2** [comment]
LGTM but needs test changes accepted.

**Comment by jbj** [comment]
Here's an additional improvement we're getting from this PR:
```diff
--- a/cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/tainted/IntegerOverflowTainted.expected
+++ b/cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/tainted/IntegerOverflowTainted.expected
@@ -1,3 +1,4 @@
 | test3.c:12:31:12:34 | * ... | $@ flows to here and is used in an expression which might overflow negatively. | test3.c:11:15:11:18 | argv | User-provided value |
 | test3.c:13:16:13:19 | * ... | $@ flows to here and is used in an expression which might overflow negatively. | test3.c:11:15:11:18 | argv | User-provided value |
 | test4.cpp:13:17:13:20 | access to array | $@ flows to here and is used in an expression which might overflow negatively. | test4.cpp:9:13:9:16 | argv | User-provided value |
+| test5.cpp:10:9:10:15 | call to strtoul | $@ flows to here and is used in an expression which might overflow. | test5.cpp:9:7:9:9 | buf | User-provided value |
```

There are also improvements in five `expected` files in the internal repo.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Pure.qll [hunk 1]
```diff
@@ -83,6 +83,11 @@ class PureStrFunction extends AliasFunction, ArrayFunction, TaintFunction, SideE
   override predicate hasOnlySpecificReadSideEffects() { none() }
 
   override predicate hasOnlySpecificWriteSideEffects() { any() }
+
+  override predicate hasSpecificReadSideEffect(ParameterIndex i, boolean buffer) {
+    getParameter(i).getUnspecifiedType() instanceof PointerType and
+    buffer = true
+  }
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Pure.qll]: If I'm understanding correctly, the information encoded here is exactly the same as the `hasArrayInput` predicate for the `ArrayFunction` model.  Is there a way we could safely have one deduce from the other?

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Pure.qll]: I've put it on our team meeting agenda to discuss how we can make it easier to get this modelling right in the future. Unfortunately I don't expect we'll have time to discuss it this week.

Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Pure.qll]: I think it will also be tricky to get right, to be honest.

Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Pure.qll]: We had a brief discussion about this after the meeting, one idea that came up was to have `ArrayFunction` model only the buffer sizes, not their input / output status - and point people to `SideEffectFunction` for that (which possibly needs explaining a bit more clearly as the read side effects in particular are not actually 'side effects' in the traditional sense).

**Pull request merged**